File: runelite-client/src/main/java/net/runelite/client/plugins/alfred/api/rs/walk/PathWalker.java
package net.runelite.client.plugins.alfred.api.rs.walk;

import net.runelite.api.GameObject;
import net.runelite.api.Point;
import net.runelite.api.Tile;
import net.runelite.api.WallObject;
import net.runelite.api.coords.WorldPoint;
import net.runelite.client.plugins.alfred.Alfred;
import net.runelite.client.plugins.alfred.api.rs.math.Calculations;
import net.runelite.client.plugins.alfred.api.rs.menu.RSMenu;
import net.runelite.client.plugins.alfred.api.rs.player.RSPlayer;
import net.runelite.client.plugins.alfred.api.rs.walk.astar.AStarNode;

import java.util.ArrayList;
import java.util.List;

public class PathWalker {
    private final List<AStarNode> tiles;
    private final RSPlayer player;

    public PathWalker(List<AStarNode> tiles) {
        this.tiles = tiles;
        this.player = Alfred.api.players().getLocalPlayer();
    }

    private List<Path> buildPath() {
        List<Path> paths = new ArrayList<>();
        int upperBound = tiles.size() - 1;

        AStarNode previousNode = null;
        for (int i = 0; i < tiles.size() - 1; i++) {
            AStarNode currentNode = tiles.get(i);
            AStarNode nextNode = null;

//            if (previousNode == currentNode) {
//                continue;
//            }

            if (i + 1 <= upperBound) {
                nextNode = tiles.get(i + 1);
            }

            if (paths.isEmpty()) {
                if (nextNode == null) {
                    paths.add(new Path(currentNode.getWorldLocation(), currentNode.getWorldLocation(), currentNode));
                } else {
                    paths.add(new Path(currentNode.getWorldLocation(), nextNode.getWorldLocation(), currentNode));
                }

                previousNode = currentNode;
                continue;
            }

            if (nextNode == null) {
                paths.add(new Path(currentNode.getWorldLocation(), currentNode.getWorldLocation(), currentNode));
                continue;
            }

            if (currentNode.getIsOperable()) {
                paths.add(new Path(currentNode.getWorldLocation(), nextNode.getWorldLocation(), currentNode));
                previousNode = currentNode;
                continue;
            }

            if (nextNode.getIsOperable()) {
                paths.add(new Path(currentNode.getWorldLocation(), nextNode.getWorldLocation(), currentNode));
                previousNode = currentNode;
                continue;
            }

            int distance = (int) Calculations.distanceBetweenPoints(previousNode.getWorldLocation(), currentNode.getWorldLocation());
            if (distance >= 5) {
                paths.add(new Path(currentNode.getWorldLocation(), nextNode.getWorldLocation(), currentNode));
                previousNode = currentNode;
            }
        }

        for (Path p : paths) {
            System.out.println(p.getStart() + ", " + p.getNode().getIsOperable() + ", " + p.getEnd());
        }

        return paths;

//        for (AStarNode currentNode : tiles) {
//            int distance = (int) Calculations.distanceBetweenPoints(player.getWorldLocation(), currentNode.getWorldLocation());
//            boolean isLastNode = currentNode.getWorldLocation().equals(tiles.get(tiles.size() - 1).getWorldLocation());
//
//            if (isLastNode && distance < 3) {
//                continue;
//
//            } else if (isLastNode) {
//                clickPoint(minimapPoint);
//                continue;
//            }
//
//            if (currentNode.getIsOperable()) {
//                RSTile realTile = findTile(currentNode);
//                if (realTile == null) {
//                    continue;
//                }
//
//                if (!operateOnTile(currentNode, realTile.getTile())) {
//                    continue;
//                }
//            }
//
//            if (distance >= 7) {
//                clickPointWhileRunning(minimapPoint, currentNode);
//            }
//        }
    }

//    public void walkPath() {
//        for (AStarNode currentNode : tiles) {
//            System.out.println(currentNode.getWorldLocation());
//
//            Point minimapPoint = Alfred.api.miniMap().getWorldPointToScreenPoint(currentNode.getWorldLocation());
//            if (minimapPoint == null) {
//                continue;
//            }
//
//            int distance = (int) Calculations.distanceBetweenPoints(player.getWorldLocation(), currentNode.getWorldLocation());
//            boolean isLastNode = currentNode.getWorldLocation().equals(tiles.get(tiles.size() - 1).getWorldLocation());
//
//            if (isLastNode && distance < 3) {
//                continue;
//
//            } else if (isLastNode) {
//                clickPoint(minimapPoint);
//                continue;
//            }
//
//            if (currentNode.getIsOperable()) {
//                RSTile realTile = findTile(currentNode);
//                if (realTile == null) {
//                    continue;
//                }
//
//                if (!operateOnTile(currentNode, realTile.getTile())) {
//                    continue;
//                }
//            }
//
//            if (distance >= 7) {
//                clickPointWhileRunning(minimapPoint, currentNode);
//            }
//        }
//    }
//
    public void walkPath() {
        int upperBound = tiles.size() - 1;

        AStarNode previousNode = null;
        for (int i = 0; i < tiles.size() - 1; i++) {
            AStarNode currentNode = tiles.get(i);
            AStarNode nextNode = null;

            if (i + 1 <= upperBound) {
                nextNode = tiles.get(i + 1);
            }

            if (nextNode == null) {
                Point minimapPoint = getMinimapPoint(currentNode.getWorldLocation());
                if (minimapPoint == null) {
                    continue;
                }
                clickPoint(minimapPoint);
                previousNode = currentNode;
                continue;
            }

            if (currentNode.getIsOperable()) {
                Path p = new Path(currentNode.getWorldLocation(), nextNode.getWorldLocation(), currentNode);
                p.walk();
                previousNode = currentNode;
                continue;
            }

            if (nextNode.getIsOperable()) {
                Path p = new Path(currentNode.getWorldLocation(), nextNode.getWorldLocation(), currentNode);
                p.walk();
                previousNode = currentNode;
                continue;
            }

            if (previousNode == null) {
                Point minimapPoint = getMinimapPoint(currentNode.getWorldLocation());
                if (minimapPoint == null) {
                    continue;
                }
                clickPoint(minimapPoint);
                previousNode = currentNode;
                continue;
            }

            int distance = (int) Calculations.distanceBetweenPoints(previousNode.getWorldLocation(), currentNode.getWorldLocation());
            if (distance >= 5) {
                Point minimapPoint = getMinimapPoint(currentNode.getWorldLocation());
                if (minimapPoint == null) {
                    continue;
                }
                clickPointWhileRunning(minimapPoint, currentNode);
                previousNode = currentNode;
            }
        }
    }

//    public void walkPath() {
//        List<Path> paths = buildPath();
//
//        for (Path currentPath : paths) {
//            currentPath.walk();
//        }
//    }
    private Point getMinimapPoint(WorldPoint worldPoint) {
        return Alfred.api.miniMap().getWorldPointToScreenPoint(worldPoint);
    }

    private void clickPoint(Point minimapPoint) {
        Alfred.getMouse().leftClick(minimapPoint);
        Alfred.sleep(1000);
        Alfred.sleepUntil(() -> !player.isMoving() && !player.isInteracting() && player.isIdle(), 200, 1000 * 10);
    }

    private void clickPointWhileRunning(Point minimapPoint, AStarNode node) {
        Alfred.getMouse().leftClick(minimapPoint);
        Alfred.sleep(1000);
        Alfred.sleepUntil(() -> {
            int distanceToTarget = (int) Calculations.distanceBetweenPoints(player.getWorldLocation(), node.getWorldLocation());
            boolean isStill = !player.isMoving() && !player.isInteracting() && player.isIdle();
            boolean nearTile = distanceToTarget <= 2;
            return isStill || nearTile;
        }, 200, 1000 * 30);
    }

    private boolean operateOnTile(AStarNode node, Tile tile) {
        Alfred.sleepUntil(() -> !player.isMoving() && !player.isInteracting() && player.isIdle(), 200, 1000 * 10);
        boolean success = false;
        if (node.getOperableName().equals("DOOR") || node.getOperableName().equals("GATE")) {
            success = operateOnDoor(tile);

        } else if (node.getOperableName().equals("STAIRS") || node.getOperableName().equals("STAIRCASE")) {
            success = operateOnStair(tile);
        }

        if (!success) {
            return false;
        }

        Alfred.sleep(1000);
        Alfred.sleepUntil(() -> !player.isMoving() && !player.isInteracting() && player.isIdle(), 200, 1000 * 10);
        Alfred.sleep(250, 500);
        return true;
    }

    private boolean operateOnDoor(Tile tile) {
        WallObject wallObject = tile.getWallObject();
        if (wallObject == null) {
            return false;
        }

        Alfred.getMouse().rightClick(wallObject.getConvexHull().getBounds());
        Alfred.sleep(200);

        RSMenu rsMenu = Alfred.api.menu().getMenu();
        if (rsMenu == null) {
            System.out.println("Menu is null");
            return false;
        }

        if (!rsMenu.hasAction("open")) {
            System.out.println("Menu does not contain open action");
            return true;
        }

        if (!rsMenu.clickAction("open")) {
            System.out.println("Failed to operate on tile");
            return false;
        }
        return true;
    }

    private boolean operateOnStair(Tile tile) {
        RSPlayer player = Alfred.api.players().getLocalPlayer();
        GameObject foundGameObject = null;

        for (GameObject gameObject : tile.getGameObjects()) {
            if (gameObject == null) {
                continue;
            }

            String tileName = Alfred.api.objects().getObjectIdVariableName(gameObject.getId());
            if (tileName.contains("STAIRS") || tileName.contains("STAIRCASE")) {
                foundGameObject = gameObject;
                break;
            }
        }

        if (foundGameObject == null) {
            return false;
        }

        Alfred.getMouse().rightClick(foundGameObject.getClickbox().getBounds());
        Alfred.sleep(200);

        int playerZ = player.getWorldLocation().getPlane();
        int tileZ = tile.getPlane();

        boolean goingUp = tileZ > playerZ;
        String action = goingUp ? "climb-up" : "climb-down";

        RSMenu rsMenu = Alfred.api.menu().getMenu();
        if (rsMenu == null) {
            System.out.println("Menu is null");
            return false;
        }

        if (!rsMenu.hasAction(action)) {
            System.out.println("Menu does not contain action");
            return true;
        }

        if (!rsMenu.clickAction(action)) {
            System.out.println("Failed to operate on tile");
            return false;
        }

        return Alfred.sleepUntil(() -> player.getWorldLocation().getPlane() == tileZ, 100, 5000);
    }


    private RSTile findTile(AStarNode node) {
        List<RSTile> tiles = Alfred.api.walk().getAllTiles();
        for (RSTile tile : tiles) {
            if (tile.getWorldLocation().equals(node.getWorldLocation())) {
                return tile;
            }
        }
        return null;
    }

    private Tile getFirstOperableNeighbor(Tile tile) {
        List<RSTile> walkableTiles = Alfred.api.walk().getWalkableTiles();

        for (int x = -1; x < 2; x++) {
            for (int y = -1; y < 2; y++) {
                if (x == 0 && y == 0) {
                    continue;
                }

                int checkX = tile.getWorldLocation().getX() + x;
                int checkY = tile.getWorldLocation().getY() + y;

                for (RSTile rsTile : walkableTiles) {
                    if (rsTile.getWorldLocation().getX() != checkX || rsTile.getWorldLocation().getY() != checkY) {
                        continue;
                    }

                    WallObject wallObject = rsTile.getTile().getWallObject();
                    if (wallObject == null) {
                        continue;
                    }

                    if (RSWalkHelper.getOperableObjectIds().contains(wallObject.getId())) {
                        return rsTile.getTile();
                    }
                }
            }
        }
        return null;
    }
}

File: runelite-client/src/main/java/net/runelite/client/plugins/alfred/api/rs/walk/pathfinder/PathWalker.kt
package net.runelite.client.plugins.alfred.api.rs.walk.pathfinder

import net.runelite.api.Point
import net.runelite.api.Tile
import net.runelite.api.coords.WorldPoint
import net.runelite.client.plugins.alfred.Alfred
import net.runelite.client.plugins.alfred.api.rs.math.Calculations
import net.runelite.client.plugins.alfred.api.rs.objects.RSObject
import net.runelite.client.plugins.alfred.api.rs.player.RSPlayer
import net.runelite.client.plugins.alfred.util.Utility

class PathWalker(private val nodes: List<PathNode>) {
    private var player: RSPlayer = Alfred.api.players.localPlayer

    fun walkPath() {
        val skipDistance = 4
        val upperBound = nodes.size - 1
        var previousNode: PathNode? = null
        var nextNode: PathNode? = null

        val player = Alfred.api.players.localPlayer

        for (currentNode in nodes) {
            val index = nodes.indexOf(currentNode)
            val isLastNode = nodes.indexOf(currentNode) == nodes.lastIndex


            Utility.retryFunction(3, true) {
                if (player.runEnergy >= 30 && !player.isRunningActive) {
                    player.toggleRunning(true)
                }
                return@retryFunction true
            }

            if (index + 1 <= upperBound) {
                nextNode = nodes.get(index + 1)
            }

            if (isLastNode) {
                val minimapPoint = getMinimapPoint(currentNode.worldLocation) ?: continue
                clickPoint(minimapPoint)
                previousNode = currentNode
                continue
            }

            if (currentNode.pathTransports.isNotEmpty()) {
                if (previousNode != null) {
                    val minimapPoint = getMinimapPoint(previousNode.worldLocation)
                    if (minimapPoint != null) {
                        clickPoint(minimapPoint)
                    }
                }

                nextNode ?: continue
                handleTransport(currentNode, nextNode)
                previousNode = currentNode
                continue
            }

            if (previousNode == null && nodes.count() < skipDistance) {
                val minimapPoint = getMinimapPoint(currentNode.worldLocation) ?: continue
                clickPoint(minimapPoint)
                previousNode = currentNode
                continue

            } else if (previousNode == null) {
                previousNode = currentNode
                continue
            }

            val distance = Calculations.distanceBetweenPoints(previousNode.worldLocation, currentNode.worldLocation).toInt()
            if (distance >= skipDistance) {
                val minimapPoint = getMinimapPoint(currentNode.worldLocation) ?: continue
                clickPointWhileRunning(minimapPoint, currentNode)
                previousNode = currentNode
            }
        }
    }

    private fun getMinimapPoint(worldPoint: WorldPoint): Point? {
        return Alfred.api.miniMap.getWorldPointToScreenPoint(worldPoint)
    }

    private fun clickPoint(minimapPoint: Point) {
        Alfred.mouse.leftClick(minimapPoint)
        Alfred.sleep(1000)
        Alfred.sleepUntil({ !player.isMoving && !player.isInteracting && player.isIdle }, 200, 1000 * 10)
    }

    private fun clickPointWhileRunning(minimapPoint: Point, node: PathNode) {
        Alfred.mouse.leftClick(minimapPoint)
        Alfred.sleep(1000)
        Alfred.sleepUntil({
            val distanceToTarget = Calculations.distanceBetweenPoints(player.worldLocation, node.worldLocation).toInt()
            val isStill = !player.isMoving && !player.isInteracting && player.isIdle
            val nearTile = distanceToTarget <= 2
            isStill || nearTile
        }, 200, 1000 * 30)
    }

    private fun handleTransport(pathNode: PathNode, nextNode: PathNode): Boolean {
        Alfred.sleepUntil({ !player.isMoving && !player.isInteracting && player.isIdle }, 200, 1000 * 10)
        Alfred.sleep(100, 250)

        val transport: PathTransport? = findTransport(pathNode, nextNode)
        if (transport == null) {
            println("cant find transport")
            return false
        }

        val success = operateTransport(transport)
        if (!success) {
            return false
        }

        Alfred.sleep(1000)
        Alfred.sleepUntil({ !player.isMoving && !player.isInteracting && player.isIdle }, 200, 1000 * 10)
        Alfred.sleep(250, 500)
        return true
    }

    private fun findTransport(pathNode: PathNode, nextNode: PathNode): PathTransport? {
        return if (pathNode.pathTransports.count() == 1) {
            pathNode.pathTransports.first()
        } else {
            pathNode.pathTransports.firstOrNull { pathTransport: PathTransport -> pathTransport.endPathNode?.worldLocation == nextNode.worldLocation }
        }
    }

    private fun operateTransport(pathTransport: PathTransport): Boolean {
//        val transportObject = Alfred.api.objects.objectsFromTiles.firstOrNull { rsObject: RSObject -> rsObject.id == pathTransport.objectId && rsObject.worldLocation == pathTransport.startPathNode.worldLocation }
//        val transportObject = Alfred.api.objects.objectsFromTiles.firstOrNull { rsObject: RSObject -> rsObject.id == pathTransport.objectId }
        val transportObject = Alfred.api.objects.objectsFromTiles
            .filter { rsObject: RSObject -> rsObject.worldLocation != null }
            .filter { rsObject: RSObject -> rsObject.id == pathTransport.objectId }
            .minBy { rsObject: RSObject -> rsObject.worldLocation!!.distanceTo(player.worldLocation) }

        if (transportObject == null) {
            println("No transport: ${pathTransport.name}, found with ID: ${pathTransport.objectId}, at location: ${pathTransport.startPathNode.worldLocation}")
            return false
        }

        println("Operating on Transport: ${pathTransport.name}, with action: ${pathTransport.action}")

        when (pathTransport.objectName.lowercase()) {
            "door" -> Alfred.mouse.rightClick(transportObject.convexHull?.bounds)
            "gate" -> Alfred.mouse.rightClick(transportObject.convexHull?.bounds)
            else -> Alfred.mouse.rightClick(transportObject.clickBox)
        }

        Alfred.sleep(200)

        if (!Alfred.sleepUntil({ Alfred.api.menu.menu.hasAction(pathTransport.action) }, 200, 2000)) {
            println("No menu action: ${pathTransport.action} found")
            return false
        }

        val rsMenu = Alfred.api.menu.menu
        if (!rsMenu.clickAction(pathTransport.action)) {
            println("Failed to click transport")
            return false
        }

        if (pathTransport.endPathNode != null) {
            return Alfred.sleepUntil({ player.worldLocation == pathTransport.endPathNode.worldLocation }, 100, 5000)
        }

        return true
    }


    private fun findTile(pathNode: PathNode): Tile? {
        val tiles = Alfred.api.walk.getAllTiles()
        for (tile in tiles) {
            if (tile.worldLocation == pathNode.worldLocation) {
                return tile.tile
            }
        }
        return null
    }
}